### PR TITLE
ExtractFeatures supports -XL exclusion and follows intervals, other optimizations

### PR DESCRIPTION
**Pull request by mmorgantaylor**
addresses specops issues:
- #268 https://github.com/broadinstitute/dsp-spec-ops/issues/268
- #270 https://github.com/broadinstitute/dsp-spec-ops/issues/270

similar changes as the PR for ExtractCohort:
- added custom classes `ExtractFeaturesRecord` that implements `Locatable`
- refactored attribute building from these records
- now that the records are `Locatable`s, can use `OverlapDetector` to filter locations down to only desired intervals (including excluded sites)

also:
- enable using intervals input (-L) rather than specifying min-location and max-location. updated WDL to support scattering using SplitIntervals (based off of CohortExtract)
- add back AS_QD to headers (currently headers are shared between ExtractCohort and ExtractFeatures - AS_QD not needed for cohort but is needed for features)

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesEngine.java [hunk 1]
```diff
@@ -128,61 +131,63 @@ private void createVQSRInputFromTableResult(final GATKAvroReader avroReader) {
...
 
-        String contig = SchemaUtils.decodeContig(location);
-        int position = SchemaUtils.decodePosition(location);
+        String contig = rec.getContig();
+        int position = rec.getStart();
         // I don't understand why the other modes iterate through a list of all columns, and then
         //  switch based on the column name rather than just getting the columns desired directly (like I'm going to do here).
         // It might be because those field names are just prefixed versions of standard VCF fields?
 
         // TODO: de-python names (no  _)
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesEngine.java:156]: I think this is done, can you remove the TODO?

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
+package org.broadinstitute.hellbender.tools.variantdb.nextgen;
+
+        import htsjdk.samtools.util.Locatable;
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:3]: fix indenting (0 spaces)

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    private final int numHomvarSamples;
+
+    // FEATURE_EXTRACT_FIELDS = Arrays.asList(
+    //      LOCATION_FIELD_NAME,
+    //      REF_ALLELE_FIELD_NAME,
+    //      "allele",
+    //      RAW_QUAL,
+    //      "ref_ad",
+    //      AS_MQRankSum,
+    //      "AS_MQRankSum_ft",
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:40]: These are the frequency tables to be able to merge annotations... if they are in the SQL query, let's keep them here as well

**Review by kcibul — APPROVED** [review]
Looks awesome, just minor comments.  Also, would be great to run this against the tieout set and make sure nothing blows up (ie the 37 samples)

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java (hunk 3 above)
Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:40]: should I do the work of parsing the frequency tables into a usable format in java, or just leave them as strings? if the former, do you know which object type I should use?

Comment (kcibul) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:40]: Just leave as a string for now

Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:40]: oh thank goodness

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java (hunk 2 above)
Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java]: thank you!

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    //      "num_homvar_samples");
+
+
+    public ExtractFeaturesRecord(GenericRecord genericRecord) {
+        this.location = Long.parseLong(genericRecord.get(SchemaUtils.LOCATION_FIELD_NAME).toString());
+        this.contig = SchemaUtils.decodeContig(location);
+        this.start = SchemaUtils.decodePosition(location);
+        this.end = start;
+
+        this.ref = Objects.toString(genericRecord.get("ref"), null);
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:63]: @kcibul should I add all these strings to SchemaUtils instead?

Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractFeaturesRecord.java:63]: i'm going with "not in this PR"

File: .dockstore.yml [hunk 5]
```diff
@@ -96,6 +96,7 @@ workflows:
        branches:
          - master
          - ah_var_store
+         - mmt_extract_features_records
```
Comment (kcibul) [inline .dockstore.yml:99]: remove before merging

**Review by kcibul — APPROVED** [review]
❤️

**Pull request merged**